### ISSUE TITLE
第10章 actor-critic算法的顺序问题

### ISSUE THREAD

**Issue by ppap36** [boyu-ai/Hands-on-RL#92]
我在复现第10章代码时，观察到update函数有如下内容
```
       td_target = rewards + self.gamma * self.critic(next_states) * (1 -  dones)
       td_delta = td_target - self.critic(states)  # 时序差分误差
       log_probs = torch.log(self.actor(states).gather(1, actions))
       actor_loss = torch.mean(-log_probs * td_delta.detach())
       # 均方误差损失函数
       critic_loss = torch.mean(F.mse_loss(self.critic(states), td_target.detach()))
       self.actor_optimizer.zero_grad()
       self.critic_optimizer.zero_grad()
       actor_loss.backward()  # 计算策略网络的梯度
       critic_loss.backward()  # 计算价值网络的梯度
       self.actor_optimizer.step()  # 更新策略网络的参数
       self.critic_optimizer.step()  # 更新价值网络的参数
```

突发奇想，如果把上面的代码整理一下，actor更新放在一起，critic更新放在一起，那么是不是结构更清晰一些，之后修改成如下内容：
```
       #更新critic网络
       td_target = rewards + self.gamma * self.critic(next_states) * (1 - dones)
       critic_loss = torch.mean(F.mse_loss(self.critic(states), td_target.detach()))
       self.critic_optimizer.zero_grad()
       critic_loss.backward()
       self.critic_optimizer.step()
       #更新actor网络
       td_delta = td_target - self.critic(states)
       log_probs = torch.log(self.actor(states).gather(1, actions))
       actor_loss = torch.mean(-log_probs * td_delta.detach())
       actor_loss.backward()
       self.actor_optimizer.step()
```
但是训练之后，结果如下
![image](https://github.com/user-attachments/assets/4e92e8fb-5334-4ce4-92a4-0ec7e6e274a4)
![image](https://github.com/user-attachments/assets/41ef5f89-17a1-4036-a1af-11eb054b5407)

本人没有神经网络的基础，请问大佬们修改代码顺序为什么会导致这样的结果？
感激不尽

**Comment by ppap36** [comment]
可能对原因了解了一二，问题出现在`detach`这个函数中，当一个变量x被执行x.detach()后，那么这个变量就不会再参与NN计算图的更新了，所以如果我修改了顺序，就会导致NN的参数在某些变量本应该被detach的时候更新，导致错误

如果有大佬知道到底which变量，when需要detach，可否指点一二

**Comment by eddie1516** [comment]
actorloss 需要 critic的输出 你先更新了critic 输出就变了 算出来td error肯定是错的

**Comment by ppap36** [comment]
感谢
